aten/src/ATen/native/FastPathConditions.h

```diff
@@ -0,0 +1,34 @@
+#include <ATen/ATen.h>
+
+namespace at {
+namespace native {
+
+namespace {
+inline bool all_dims_are_same(const Tensor& t1, const Tensor& t2) {
+  if (t1.ndimension() != t2.ndimension()) {
+    return false;
+  }
+
+  IntArrayRef t1_size = t1.sizes();
+  IntArrayRef t2_size = t2.sizes();
+  for (int64_t i = 0; i < t1.ndimension(); ++i) {
+    if (t1_size[i] != t2_size[i]) {
+      return false;
+    }
+  }
+  return true;
+}
+} // namespace
+
+namespace binary_op_fast_path_conditions {
+
+bool enable_contiguous_eq_size_cpu_fastpath(const Tensor& t1, const Tensor& t2) {
+  return all_dims_are_same(t1, t2) && t2.is_contiguous() &&
+         t1.device().is_cpu() && t2.device().is_cpu() &&
+         t1.is_contiguous() &&
```

Comment:
I'd move the `is_cpu` checks to the front, because they're likely to be cheap and can let us quickly move on to the generic path on other devices.

Reply:
@apaszke, I actually had it that way before, but I was trying to reduce the overhead of fastpath condition checking. Since we want fastpath for inference where it is likely that the tensors are on cpu most likely those conditions will hold true. Thus I moved the condition which are less likely to be true to the front. I havent gotten around to digging into what is contributing to the overhead of condition checking, when we dont go through the fast path. So this is one place I will explore to see how to get out of here quickly.

Reply:
This is not what you should optimize for. If your assumption is that you can use the fast path, then the order of those checks is actually irrelevant, because they will all have to be executed anyway. What's important is which conditions are the ones which will fail the fastest if you _can't_ use the fast path (because then you'll get the short-circuit very quickly). Hence, it's better to put the cheap and quickly failing ones first.

I understand the desire to make inference fast, but you have to remember that this library has a thousand other users with other constraints.